### DISCUSSION TITLE
Reflection for External Crates

### DISCUSSION THREAD

**Discussion by jnhyatt** [bevyengine/bevy#8349] [Q&A] [answered]
I've got a component I'm trying to reflect (I want to modify its contents at runtime through `bevy_editor_pls`), but `#[derive(Reflect)]` fails because the struct has a `chrono::DateTime<Utc>` member. I can't manually implement `Reflect` for it because of Rust's coherence rules. Here are the options I see:

1. Replace usages of `chrono` types in my struct with something that can be reflected
2. Derive `Reflect` in the `chrono` crate and lock it behind a `bevy_reflect` feature and pull request
3. Derive `Reflect` in the `bevy_reflect` crate and lock it behind a `chrono` feature and pull request
4. Ask for guidance on Bevy's GitHub discussions page

Options 1-3 seem overly intrusive:

1. I shouldn't have to sacrifice features for functionality
2. Bevy is unrelated to `chrono`
3. Bevy does not depend on `chrono`

So I opted for option 4 in the hopes that Bevy has some other way to do this that I haven't yet happened upon. Any advice would be appreciated!

**Comment by nicopap** [comment]
You can't currently reflect a foreign type, you would need to `#[reflect(ignore)]` it. Although there is some work toward it, as seen in <https://github.com/bevyengine/bevy/pull/6042>.

A workaround is possible using the [bevy_mod_component_mirror](https://lib.rs/crates/bevy_mod_component_mirror) crate (disclaimer: am the author)

**Comment by jnhyatt** [comment]
Thanks, not sure why my searching didn't turn up that issue. I'll take a look at your crate, is it exclusively for rapier?

**Reply by nicopap** [reply, marked as answer]
It's not exclusive to rapier. Only the pre-made ready-to-use stuff. For your components, you'll have to follow the instructions in the ["Implement your own mirrors" section](https://lib.rs/crates/bevy_mod_component_mirror#:~:text=Implement%20your%20own%20mirrors)

**Reply by jnhyatt** [reply]
Thanks, very much appreciated!
